From 617747450aee48e592d701de90b570b6729007ad Mon Sep 17 00:00:00 2001
From: Shunhua Lan <lsh@rock-chips.com>
Date: Wed, 11 Aug 2021 15:49:29 +0800
Subject: [PATCH 062/232] media: rk628-for-all: rk628_bt1120_v4l2: fix audio
 issues

1. reset hdmi in cru when hotplug or res change to avoid audio L/R channel swap
2. disable hdmirx and audio unit when PHY setup
3. enable audio unit after N CTS updating
4. enable audio interrupts before PHY setup
5. double init audio fifo when fifo both underflow and overflow
6. media: i2c: fix i2c access timeout
7. add 60ms delay after phy locked

Change-Id: Iec24861941bae98a04e94b1122b78f27dd3754b4
Signed-off-by: Shunhua Lan <lsh@rock-chips.com>
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628_bt1120_v4l2.c | 529 ++++++++++++++------
 1 file changed, 365 insertions(+), 164 deletions(-)

diff --git a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
index 9765585f2a9bf1..6da71b523d60ed 100644
--- a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
@@ -58,7 +58,7 @@ MODULE_PARM_DESC(debug, "debug level (0-3)");
 #define POLL_INTERVAL_MS		1000
 #define MODETCLK_CNT_NUM		1000
 #define MODETCLK_HZ			49500000
-#define RXPHY_CFG_MAX_TIMES		3
+#define RXPHY_CFG_MAX_TIMES		15
 #define CSITX_ERR_RETRY_TIMES		3
 
 #define USE_4_LANES			4
@@ -66,6 +66,27 @@ MODULE_PARM_DESC(debug, "debug level (0-3)");
 /* Test Code: 0x44 (HS RX Control of Lane 0) */
 #define HSFREQRANGE(x)			UPDATE(x, 6, 1)
 
+#define INIT_FIFO_STATE			64
+
+struct rk628_audiostate {
+	u32 hdmirx_aud_clkrate;
+	u32 fs_audio;
+	u32 ctsn_flag;
+	u32 fifo_flag;
+	int init_state;
+	int pre_state;
+	bool fifo_int;
+};
+
+#define is_validfs(x) (x == 32000 || \
+			x == 44100 || \
+			x == 48000 || \
+			x == 88200 || \
+			x == 96000 || \
+			x == 176400 || \
+			x == 192000 || \
+			x == 768000)
+
 struct rk628_bt1120 {
 	struct device *dev;
 	struct i2c_client *i2c_client;
@@ -88,6 +109,7 @@ struct rk628_bt1120 {
 	struct clk *clk_rx_read;
 	struct delayed_work delayed_work_enable_hotplug;
 	struct delayed_work delayed_work_res_change;
+	struct delayed_work delayed_work_audio_rate_change;
 	struct delayed_work delayed_work_audio;
 	struct timer_list timer;
 	struct work_struct work_i2c_poll;
@@ -101,7 +123,6 @@ struct rk628_bt1120 {
 	u64 lane_mbps;
 	u8 bt1120_lanes_in_use;
 	u32 mbus_fmt_code;
-	u64 fs_audio;
 	u8 fps;
 	u32 stream_state;
 	int hdmirx_irq;
@@ -112,8 +133,12 @@ struct rk628_bt1120 {
 	bool audio_present;
 	bool hpd_output_inverted;
 	bool avi_rcv_rdy;
+	bool ctsn_ints_en;
+	bool vid_ints_en;
+	bool fifo_ints_en;
 	bool dual_edge;
 	struct rk628_hdcp hdcp;
+	struct rk628_audiostate audio_state;
 };
 
 struct rk628_bt1120_mode {
@@ -226,13 +251,14 @@ static int rk628_bt1120_s_dv_timings(struct v4l2_subdev *sd,
 static int rk628_bt1120_s_edid(struct v4l2_subdev *sd,
 				struct v4l2_subdev_edid *edid);
 static int rk628_hdmirx_phy_power_on(struct v4l2_subdev *sd);
+static int rk628_hdmirx_phy_power_off(struct v4l2_subdev *sd);
 static int rk628_hdmirx_phy_setup(struct v4l2_subdev *sd);
 static void rk628_bt1120_format_change(struct v4l2_subdev *sd);
 static void enable_stream(struct v4l2_subdev *sd, bool enable);
 static void rk628_hdmirx_audio_setup(struct v4l2_subdev *sd);
 static void rk628_hdmirx_vid_enable(struct v4l2_subdev *sd, bool en);
 static void rk628_hdmirx_hpd_ctrl(struct v4l2_subdev *sd, bool en);
-
+static void rk628_hdmirx_controller_reset(struct v4l2_subdev *sd);
 static inline struct rk628_bt1120 *to_bt1120(struct v4l2_subdev *sd)
 {
 	return container_of(sd, struct rk628_bt1120, sd);
@@ -245,14 +271,14 @@ static bool tx_5v_power_present(struct v4l2_subdev *sd)
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
 
 	cnt = 0;
-	for (i = 0; i < 10; i++) {
+	for (i = 0; i < 5; i++) {
 		val = gpiod_get_value(bt1120->plugin_det_gpio);
 		if (val > 0)
 			cnt++;
-		usleep_range(2000, 2100);
+		usleep_range(500, 600);
 	}
 
-	ret = (cnt >= 7) ? true : false;
+	ret = (cnt >= 3) ? true : false;
 	v4l2_dbg(1, debug, sd, "%s: %d\n", __func__, ret);
 
 	return ret;
@@ -280,9 +306,27 @@ static int get_audio_sampling_rate(struct v4l2_subdev *sd)
 	if (no_signal(sd))
 		return 0;
 
-	return bt1120->fs_audio;
+	return bt1120->audio_state.fs_audio;
 }
 
+static void rk628_hdmirx_ctrl_enable(struct v4l2_subdev *sd, int en)
+{
+	u32 mask;
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	if (en) {
+		/* don't enable audio until N CTS updated */
+		mask = HDMI_ENABLE_MASK;
+		v4l2_dbg(1, debug, sd, "%s: %#x %d\n", __func__, mask, en);
+		rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF,
+				   mask, HDMI_ENABLE(1) | AUD_ENABLE(1));
+	} else {
+		mask = AUD_ENABLE_MASK | HDMI_ENABLE_MASK;
+		v4l2_dbg(1, debug, sd, "%s: %#x %d\n", __func__, mask, en);
+		rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF,
+				   mask, HDMI_ENABLE(0) | AUD_ENABLE(0));
+	}
+}
 static int rk628_bt1120_get_detected_timings(struct v4l2_subdev *sd,
 				     struct v4l2_dv_timings *timings)
 {
@@ -412,22 +456,22 @@ static void rk628_bt1120_delayed_work_enable_hotplug(struct work_struct *work)
 	v4l2_dbg(1, debug, sd, "%s: 5v_det:%d\n", __func__, plugin);
 	if (plugin) {
 		rk628_bt1120_enable_interrupts(sd, false);
+		rk628_hdmirx_audio_setup(sd);
 		rk628_hdmirx_set_hdcp(bt1120->rk628, &bt1120->hdcp, bt1120->enable_hdcp);
 		rk628_hdmirx_hpd_ctrl(sd, true);
 		rk628_hdmirx_config_all(sd);
 		rk628_bt1120_enable_interrupts(sd, true);
 		rk628_i2c_update_bits(bt1120->rk628, GRF_SYSTEM_CON0,
 				SW_I2S_DATA_OEN_MASK, SW_I2S_DATA_OEN(0));
-		schedule_delayed_work(&bt1120->delayed_work_audio, 0);
 	} else {
 		rk628_bt1120_enable_interrupts(sd, false);
 		enable_stream(sd, false);
 		cancel_delayed_work(&bt1120->delayed_work_res_change);
+		cancel_delayed_work(&bt1120->delayed_work_audio);
 		rk628_hdmirx_hpd_ctrl(sd, false);
+		rk628_hdmirx_phy_power_off(sd);
+		rk628_hdmirx_controller_reset(sd);
 		bt1120->nosignal = true;
-		rk628_i2c_update_bits(bt1120->rk628, GRF_SYSTEM_CON0,
-				SW_I2S_DATA_OEN_MASK, SW_I2S_DATA_OEN(1));
-		cancel_delayed_work(&bt1120->delayed_work_audio);
 	}
 	mutex_unlock(&bt1120->confctl_mutex);
 }
@@ -457,7 +501,6 @@ static int rk628_check_resulotion_change(struct v4l2_subdev *sd)
 	return 0;
 }
 
-
 static void rk628_delayed_work_res_change(struct work_struct *work)
 {
 	struct delayed_work *dwork = to_delayed_work(work);
@@ -473,6 +516,12 @@ static void rk628_delayed_work_res_change(struct work_struct *work)
 	if (plugin) {
 		if (rk628_check_resulotion_change(sd)) {
 			v4l2_dbg(1, debug, sd, "res change, recfg ctrler and phy!\n");
+			cancel_delayed_work_sync(&bt1120->delayed_work_audio);
+			rk628_hdmirx_phy_power_off(sd);
+			rk628_hdmirx_controller_reset(sd);
+			rk628_hdmirx_audio_setup(sd);
+			rk628_hdmirx_set_hdcp(bt1120->rk628, &bt1120->hdcp, bt1120->enable_hdcp);
+			rk628_hdmirx_hpd_ctrl(sd, true);
 			rk628_hdmirx_config_all(sd);
 			rk628_bt1120_enable_interrupts(sd, true);
 			rk628_i2c_update_bits(bt1120->rk628, GRF_SYSTEM_CON0,
@@ -860,6 +909,7 @@ static int rk628_hdmirx_phy_power_on(struct v4l2_subdev *sd)
 	}
 
 	if (bt1120->rxphy_pwron == false) {
+		rk628_hdmirx_ctrl_enable(sd, 0);
 		ret = rk628_rxphy_power_on(bt1120->rk628, f);
 		if (ret) {
 			bt1120->rxphy_pwron = false;
@@ -867,28 +917,36 @@ static int rk628_hdmirx_phy_power_on(struct v4l2_subdev *sd)
 		} else {
 			bt1120->rxphy_pwron = true;
 		}
+		rk628_hdmirx_ctrl_enable(sd, 1);
+		msleep(100);
 	}
 
 	return ret;
 }
 
+static int rk628_hdmirx_phy_power_off(struct v4l2_subdev *sd)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	if (bt1120->rxphy_pwron) {
+		v4l2_dbg(1, debug, sd, "rxphy power off!\n");
+		rk628_rxphy_power_off(bt1120->rk628);
+		bt1120->rxphy_pwron = false;
+	}
+	usleep_range(100, 100);
+
+	return 0;
+}
+
 static void rk628_hdmirx_vid_enable(struct v4l2_subdev *sd, bool en)
 {
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
 
 	v4l2_dbg(1, debug, sd, "%s: %sable\n", __func__, en ? "en" : "dis");
 	if (en) {
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_SAO_CTRL,
-				I2S_LPCM_BPCUV(0) |
-				I2S_32_16(1)|
-				I2S_ENABLE_BITS(0));
 		rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF,
 				VID_ENABLE_MASK, VID_ENABLE(1));
 	} else {
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_SAO_CTRL,
-				I2S_LPCM_BPCUV(0) |
-				I2S_32_16(1)|
-				I2S_ENABLE_BITS(0x3f));
 		rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF,
 				VID_ENABLE_MASK, VID_ENABLE(0));
 	}
@@ -908,6 +966,97 @@ static void rk628_hdmirx_controller_reset(struct v4l2_subdev *sd)
 	rk628_i2c_write(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF, 0x0001017f);
 }
 
+static void rk628_hdmirx_audio_fifo_init(struct v4l2_subdev *sd)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	v4l2_dbg(1, debug, sd, "%s\n", __func__);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_ICLR, 0x1f);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10001);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10000);
+	bt1120->audio_state.pre_state = bt1120->audio_state.init_state = INIT_FIFO_STATE*4;
+}
+
+static void rk628_hdmirx_audio_fifo_initd(struct v4l2_subdev *sd)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	v4l2_dbg(1, debug, sd, "%s\n", __func__);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_ICLR, 0x1f);
+	rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_AUD_FIFO_TH,
+			   AFIF_TH_START_MASK,
+			   AFIF_TH_START(192));
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10001);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10000);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10001);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10000);
+	rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_AUD_FIFO_TH,
+			   AFIF_TH_START_MASK,
+			   AFIF_TH_START(INIT_FIFO_STATE));
+	bt1120->audio_state.pre_state = bt1120->audio_state.init_state = INIT_FIFO_STATE*4;
+}
+
+static u32 rk628_hdmirx_audio_fs(struct v4l2_subdev *sd)
+{
+	u64 tmdsclk = 0;
+	u32 clkrate = 0, cts_decoded = 0, n_decoded = 0, fs_audio = 0;
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	/* fout=128*fs=ftmds*N/CTS */
+	rk628_i2c_read(bt1120->rk628, HDMI_RX_HDMI_CKM_RESULT, &clkrate);
+	clkrate = clkrate & 0xffff;
+	/* tmdsclk = (clkrate/1000) * 49500000 */
+	tmdsclk = clkrate * (49500000 / 1000);
+	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_ACR_CTS, &cts_decoded);
+	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_ACR_N, &n_decoded);
+	if (cts_decoded != 0) {
+		fs_audio = div_u64((tmdsclk * n_decoded), cts_decoded);
+		fs_audio /= 128;
+		fs_audio = div_u64(fs_audio + 50, 100);
+		fs_audio *= 100;
+	}
+	v4l2_dbg(1, debug, sd,
+		 "%s: clkrate:%u tmdsclk:%llu, n_decoded:%u, cts_decoded:%u, fs_audio:%u\n",
+		 __func__, clkrate, tmdsclk, n_decoded, cts_decoded, fs_audio);
+	if (!is_validfs(fs_audio))
+		fs_audio = 0;
+	return fs_audio;
+}
+
+static void rk628_hdmirx_audio_clk_set_rate(struct v4l2_subdev *sd, u32 rate)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+
+	v4l2_dbg(2, debug, sd, "%s: %u to %u\n",
+		 __func__, bt1120->audio_state.hdmirx_aud_clkrate, rate);
+	rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, rate);
+	bt1120->audio_state.hdmirx_aud_clkrate = rate;
+}
+
+static void rk628_hdmirx_audio_clk_inc_rate(struct v4l2_subdev *sd, int dis)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+	u32 hdmirx_aud_clkrate = bt1120->audio_state.hdmirx_aud_clkrate + dis;
+
+	v4l2_dbg(2, debug, sd, "%s: %u to %u\n",
+		 __func__, bt1120->audio_state.hdmirx_aud_clkrate, hdmirx_aud_clkrate);
+	rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, hdmirx_aud_clkrate);
+	bt1120->audio_state.hdmirx_aud_clkrate = hdmirx_aud_clkrate;
+}
+
+static void rk628_hdmirx_audio_set_fs(struct v4l2_subdev *sd, u32 fs_audio)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+	u32 hdmirx_aud_clkrate_t = fs_audio*128;
+
+	v4l2_dbg(2, debug, sd, "%s: %u to %u with fs %u\n", __func__,
+		 bt1120->audio_state.hdmirx_aud_clkrate, hdmirx_aud_clkrate_t,
+		 fs_audio);
+	rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, hdmirx_aud_clkrate_t);
+	bt1120->audio_state.hdmirx_aud_clkrate = hdmirx_aud_clkrate_t;
+	bt1120->audio_state.fs_audio = fs_audio;
+}
+
 static void rk628_hdmirx_audio_setup(struct v4l2_subdev *sd)
 {
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
@@ -915,9 +1064,15 @@ static void rk628_hdmirx_audio_setup(struct v4l2_subdev *sd)
 
 	audio_pll_n = 5644;
 	audio_pll_cts = 148500;
-
-	rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, 5644800);
-
+	bt1120->audio_state.ctsn_flag = 0;
+	bt1120->audio_state.fs_audio = 0;
+	bt1120->audio_state.pre_state = 0;
+	bt1120->audio_state.init_state = INIT_FIFO_STATE*4;
+	bt1120->audio_state.fifo_int = false;
+	bt1120->fifo_ints_en = false;
+	bt1120->ctsn_ints_en = false;
+
+	rk628_hdmirx_audio_clk_set_rate(sd, 5644800);
 	/* manual aud CTS */
 	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUDPLL_GEN_CTS, audio_pll_cts);
 	/* manual aud N */
@@ -953,117 +1108,105 @@ static void rk628_hdmirx_audio_setup(struct v4l2_subdev *sd)
 			AUTO_ACLK_MUTE(2)	|
 			AUD_MUTE_SPEED(1)	|
 			AUD_AVMUTE_EN(1)	|
-			AUD_MUTE_SEL(1)		|
+			AUD_MUTE_SEL(0)		|
 			AUD_MUTE_MODE(1));
 
 	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_PAO_CTRL,
 			PAO_RATE(0));
 	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_CHEXTR_CTRL,
 			AUD_LAYOUT_CTRL(1));
-
+	bt1120->ctsn_ints_en = true;
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_SET, ACR_N_CHG_ICLR | ACR_CTS_CHG_ICLR);
 	/* audio detect */
 	rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_AUDIODET_CTRL,
 			AUDIODET_THRESHOLD(0));
 }
 
-static void rk628_bt1120_delayed_work_audio(struct work_struct *work)
+static void rk628_csi_delayed_work_audio(struct work_struct *work)
 {
 	struct delayed_work *dwork = to_delayed_work(work);
 	struct rk628_bt1120 *bt1120 = container_of(dwork, struct rk628_bt1120,
-			delayed_work_audio);
-	struct v4l2_subdev *sd = &bt1120->sd;
-	u32 val;
-	static int hdmirx_aud_clkrate = 5644800,
-		   init_state = 256, pre_state, cur_state;
-	u32 clkrate = 0, cts_decoded = 0, n_decoded = 0;
-	u64 tmdsclk = 0, fs_audio = 0;
-	static u64 pre_fs_audio;
-
-	/* fout=128*fs=ftmds*N/CTS */
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_HDMI_CKM_RESULT, &clkrate);
-	clkrate = clkrate & 0xffff;
-	/* tmdsclk = (clkrate/1000) * 49500000 */
-	tmdsclk = clkrate * (49500000 / 1000);
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_ACR_CTS, &cts_decoded);
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_ACR_N, &n_decoded);
-	/* fs_audio = ((tmdsclk * n_decoded) / cts_decoded ) / 128 */
-	if (cts_decoded != 0) {
-		fs_audio = div_u64((tmdsclk * n_decoded), cts_decoded);
-		fs_audio = div_u64(fs_audio, 128);
-		fs_audio = div_u64(fs_audio + 50, 100);
-		fs_audio *= 100;
-	}
-	v4l2_dbg(2, debug, sd,
-		"%s: clkrate:%d tmdsclk:%llu, n_decoded:%d, cts_decoded:%d, fs_audio:%llu\n",
-		__func__, clkrate, tmdsclk, n_decoded, cts_decoded, fs_audio);
-	if ((fs_audio != 0) && (abs(fs_audio - pre_fs_audio) > 1000)) {
-		hdmirx_aud_clkrate = 128 * fs_audio;
-		switch (fs_audio) {
-		case 88200:
-			hdmirx_aud_clkrate = 11111000;
-			break;
-		case 96000:
-			hdmirx_aud_clkrate = 12121000;
-			break;
-		case 176400:
-			hdmirx_aud_clkrate = 22222000;
-			break;
-		case 192000:
-			hdmirx_aud_clkrate = 23529000;
-			break;
-		default:
-			break;
-		}
-		rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, hdmirx_aud_clkrate);
-		v4l2_dbg(2, debug, sd,
-			"%s: audo switch clk_hdmirx_aud to %d  fs_audio:%llu pre_fs_audio:%llu\n",
-			__func__, hdmirx_aud_clkrate, fs_audio, pre_fs_audio);
-		if (pre_fs_audio != 0) {
-			rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_ICLR, 0x1f);
-			rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10001);
-			rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10000);
-		}
-		pre_fs_audio = fs_audio;
-		bt1120->fs_audio = fs_audio;
-	}
-
+						   delayed_work_audio);
+	struct v4l2_subdev *sd = &bt1120 ->sd;
+	struct rk628_audiostate *audio_state = &bt1120->audio_state;
+	u32 fs_audio;
+	int cur_state, init_state, pre_state;
+
+	init_state = audio_state->init_state;
+	pre_state = audio_state->pre_state;
+	fs_audio = rk628_hdmirx_audio_fs(sd);
+	if (!is_validfs(fs_audio)) {
+		v4l2_dbg(2, debug, sd, "%s: no supported fs(%u)\n", __func__, fs_audio);
+		goto exit;
+	}
+	if (abs(fs_audio - audio_state->fs_audio) > 1000)
+		rk628_hdmirx_audio_set_fs(sd, fs_audio);
 	rk628_i2c_read(bt1120->rk628, HDMI_RX_AUD_FIFO_FILLSTS1, &cur_state);
-	v4l2_dbg(2, debug, sd,
-		"%s: HDMI_RX_AUD_FIFO_FILLSTS1:%#x, single offset:%d, total offset:%d\n",
-		__func__, cur_state, cur_state-pre_state, cur_state-init_state);
+	v4l2_dbg(2, debug, sd, "%s: HDMI_RX_AUD_FIFO_FILLSTS1:%#x, single offset:%d, total offset:%d\n",
+		 __func__, cur_state, cur_state - pre_state, cur_state - init_state);
 	if (cur_state != 0)
 		bt1120->audio_present = true;
 	else
 		bt1120->audio_present = false;
-	if ((cur_state-init_state) > 16 && (cur_state-pre_state) > 0) {
-		hdmirx_aud_clkrate += 10;
-		rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, hdmirx_aud_clkrate);
-		v4l2_dbg(2, debug, sd, "%s: (cur_state-init_state) > 16 hdmirx_aud_clkrate:%d\n",
-				__func__, hdmirx_aud_clkrate);
-	} else if ((cur_state != 0) && (cur_state-init_state) < -16 && (cur_state-pre_state) < 0) {
-		hdmirx_aud_clkrate -= 10;
-		rk628_clk_set_rate(bt1120->rk628, CGU_CLK_HDMIRX_AUD, hdmirx_aud_clkrate);
-		v4l2_dbg(2, debug, sd, "%s: (cur_state-init_state) < -16 hdmirx_aud_clkrate:%d\n",
-				__func__, hdmirx_aud_clkrate);
-	}
-	pre_state = cur_state;
-
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_AUD_FIFO_ISTS, &val);
-	v4l2_dbg(2, debug, sd, "%s: HDMI_RX_AUD_FIFO_ISTS:%#x\n", __func__, val);
-	if ((val != 0x9) && ((val & 0x10) || (val & 0x8))) {
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_ICLR, 0x1f);
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10001);
-		/*msleep(1);*/
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_CTRL, 0x10000);
-		pre_state = cur_state = 0;
-
-		v4l2_err(sd, "%s: HDMI_RX_AUD_FIFO_ISTS:%#x, underflow or overflow\n",
-				__func__, val);
-	}
 
+	if ((cur_state - init_state) > 16 && (cur_state - pre_state) > 0)
+		rk628_hdmirx_audio_clk_inc_rate(sd, 10);
+	else if ((cur_state != 0) && (cur_state - init_state) < -16 && (cur_state - pre_state) < 0)
+		rk628_hdmirx_audio_clk_inc_rate(sd, -10);
+	audio_state->pre_state = cur_state;
+exit:
 	schedule_delayed_work(&bt1120->delayed_work_audio, msecs_to_jiffies(1000));
 }
 
+static void rk628_csi_delayed_work_audio_rate_change(struct work_struct *work)
+{
+	u32 fifo_ints, fifo_fillsts;
+	u32 fs_audio;
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct rk628_bt1120 *bt1120 = container_of(dwork, struct rk628_bt1120,
+						   delayed_work_audio_rate_change);
+	struct v4l2_subdev *sd = &bt1120 ->sd;
+
+	mutex_lock(&bt1120 ->confctl_mutex);
+	fs_audio = rk628_hdmirx_audio_fs(sd);
+	v4l2_dbg(1, debug, sd, "%s get audio fs %u\n", __func__, fs_audio);
+	if (bt1120->audio_state.ctsn_flag == (ACR_N_CHG_ICLR | ACR_CTS_CHG_ICLR)) {
+		bt1120->audio_state.ctsn_flag = 0;
+		if (is_validfs(fs_audio)) {
+			rk628_hdmirx_audio_set_fs(sd, fs_audio);
+			rk628_i2c_read(bt1120->rk628, HDMI_RX_AUD_FIFO_ISTS, &fifo_ints);
+			v4l2_dbg(1, debug, sd, "%s fifo ints %#x\n", __func__, fifo_ints);
+			if ((fifo_ints & 0x18) == 0x18)
+				rk628_hdmirx_audio_fifo_initd(sd);
+			else if (fifo_ints & 0x18)
+				rk628_hdmirx_audio_fifo_init(sd);
+			rk628_i2c_update_bits(bt1120->rk628, HDMI_RX_DMI_DISABLE_IF,
+					   AUD_ENABLE_MASK, AUD_ENABLE(1));
+			/* We start audio work after recieveing cts n interrupt */
+			bt1120->fifo_ints_en = true;
+			rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_IEN_SET,
+					AFIF_OVERFL_ISTS | AFIF_UNDERFL_ISTS);
+			schedule_delayed_work(&bt1120->delayed_work_audio, msecs_to_jiffies(1000));
+		} else {
+			v4l2_warn(sd, "%s invalid fs when ctsn updating\n", __func__);
+		}
+	}
+	if (bt1120 ->audio_state.fifo_int) {
+		bt1120 ->audio_state.fifo_int = false;
+		if (is_validfs(fs_audio))
+			rk628_hdmirx_audio_set_fs(sd, fs_audio);
+		rk628_i2c_read(bt1120 ->rk628, HDMI_RX_AUD_FIFO_FILLSTS1, &fifo_fillsts);
+		if (!fifo_fillsts) {
+			v4l2_dbg(1, debug, sd, "%s underflow after overflow\n", __func__);
+			rk628_hdmirx_audio_fifo_initd(sd);
+		} else {
+			v4l2_dbg(1, debug, sd, "%s overflow after underflow\n", __func__);
+			rk628_hdmirx_audio_fifo_initd(sd);
+		}
+	}
+	mutex_unlock(&bt1120 ->confctl_mutex);
+}
+
 static bool rk628_rcv_supported_res(struct v4l2_subdev *sd, u32 width,
 		u32 height)
 {
@@ -1143,6 +1286,8 @@ static void rk628_bt1120_initial_setup(struct v4l2_subdev *sd)
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
 	struct v4l2_subdev_edid def_edid;
 
+	rk628_clk_mux_testout(bt1120->rk628, CGU_CLK_HDMIRX_AUD);
+	rk628_i2c_write(bt1120->rk628, GRF_GPIO1AB_SEL_CON, HIWORD_UPDATE(0x1, 0, 0)); /* enable gpio1a0 TESTclkout */
 	/* selete int io function */
 	rk628_i2c_write(bt1120->rk628, GRF_GPIO3AB_SEL_CON, 0x30002000);
 	rk628_i2c_write(bt1120->rk628, GRF_GPIO1AB_SEL_CON, HIWORD_UPDATE(0x7, 10, 8));
@@ -1190,7 +1335,6 @@ static void rk628_bt1120_initial_setup(struct v4l2_subdev *sd)
 	def_edid.edid = edid_init_data;
 	rk628_bt1120_s_edid(sd, &def_edid);
 	rk628_hdmirx_set_hdcp(bt1120->rk628, &bt1120->hdcp, false);
-	rk628_hdmirx_audio_setup(sd);
 
 	if (tx_5v_power_present(sd))
 		schedule_delayed_work(&bt1120->delayed_work_enable_hotplug, 1000);
@@ -1224,38 +1368,79 @@ static void rk628_bt1120_format_change(struct v4l2_subdev *sd)
 
 static void rk628_bt1120_enable_interrupts(struct v4l2_subdev *sd, bool en)
 {
-	u32 val_a, val_b;
+	u32 pdec_ien, md_ien;
+	u32 pdec_mask = 0, md_mask = 0;
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
 
+	pdec_mask |= AVI_RCV_ENSET;
+	md_mask = VACT_LIN_ENSET | HACT_PIX_ENSET | HS_CLK_ENSET |
+		  DE_ACTIVITY_ENSET | VS_ACT_ENSET | HS_ACT_ENSET;
 	v4l2_dbg(1, debug, sd, "%s: %sable\n", __func__, en ? "en" : "dis");
 	/* clr irq */
-	rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_ICLR, 0xffffffff);
-	rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_ICLR, 0xffffffff);
-
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_ICLR, md_mask);
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_ICLR, pdec_mask);
 	if (en) {
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_IEN_SET,
-				VACT_LIN_ENSET |
-				HACT_PIX_ENSET |
-				HS_CLK_ENSET |
-				DE_ACTIVITY_ENSET |
-				VS_ACT_ENSET |
-				HS_ACT_ENSET);
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_SET, AVI_RCV_ENSET);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_IEN_SET, md_mask);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_SET, pdec_mask);
+		bt1120->vid_ints_en = true;
 	} else {
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_IEN_CLR, 0xffffffff);
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_CLR, 0xffffffff);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_MD_IEN_CLR, md_mask);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_CLR, pdec_mask);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_IEN_CLR, 0x1f);
+		bt1120->vid_ints_en = false;
 	}
 	usleep_range(5000, 5000);
+	rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_IEN, &md_ien);
+	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_IEN, &pdec_ien);
+	v4l2_dbg(1, debug, sd, "%s MD_IEN:%#x, PDEC_IEN:%#x\n", __func__, md_ien, pdec_ien);
+}
 
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_IEN, &val_a);
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_IEN, &val_b);
-	v4l2_dbg(1, debug, sd, "%s MD_IEN:%#x, PDEC_IEN:%#x\n",
-			__func__, val_a, val_b);
+static void rk628_csi_isr_ctsn(struct v4l2_subdev *sd, u32 pdec_ints)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+	u32 ctsn_mask = ACR_N_CHG_ICLR | ACR_CTS_CHG_ICLR;
+
+	v4l2_dbg(1, debug, sd, "%s: pdec_ints:%#x\n", __func__, pdec_ints);
+	/* cts & n both need update but maybe come diff int */
+	if (pdec_ints & ACR_N_CHG_ICLR)
+		bt1120->audio_state.ctsn_flag |= ACR_N_CHG_ICLR;
+	if (pdec_ints & ACR_CTS_CHG_ICLR)
+		bt1120->audio_state.ctsn_flag |= ACR_CTS_CHG_ICLR;
+	if (bt1120->audio_state.ctsn_flag == ctsn_mask) {
+		v4l2_dbg(1, debug, sd, "%s: ctsn updated, disable ctsn int\n", __func__);
+		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_CLR, ctsn_mask);
+		bt1120->ctsn_ints_en = false;
+		schedule_delayed_work(&bt1120->delayed_work_audio_rate_change, 0);
+	}
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_ICLR, pdec_ints & ctsn_mask);
+}
+
+static void rk628_csi_isr_fifoints(struct v4l2_subdev *sd, u32 fifo_ints)
+{
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+	u32 fifo_mask = AFIF_OVERFL_ISTS | AFIF_UNDERFL_ISTS;
+
+	v4l2_dbg(1, debug, sd, "%s: fifo_ints:%#x\n", __func__, fifo_ints);
+	/* cts & n both need update but maybe come diff int */
+	if (fifo_ints & AFIF_OVERFL_ISTS) {
+		v4l2_dbg(1, debug, sd, "%s: Audio FIFO overflow\n", __func__);
+		bt1120->audio_state.fifo_flag |= AFIF_OVERFL_ISTS;
+	}
+	if (fifo_ints & AFIF_UNDERFL_ISTS) {
+		v4l2_dbg(1, debug, sd, "%s: Audio FIFO underflow\n", __func__);
+		bt1120->audio_state.fifo_flag |= AFIF_UNDERFL_ISTS;
+	}
+	if (bt1120->audio_state.fifo_flag == fifo_mask) {
+		bt1120->audio_state.fifo_int = true;
+		bt1120->audio_state.fifo_flag = 0;
+		schedule_delayed_work(&bt1120->delayed_work_audio_rate_change, 0);
+	}
+	rk628_i2c_write(bt1120->rk628, HDMI_RX_AUD_FIFO_ICLR, fifo_ints & fifo_mask);
 }
 
 static int rk628_bt1120_isr(struct v4l2_subdev *sd, u32 status, bool *handled)
 {
-	u32 md_ints, pdec_ints, hact, vact;
+	u32 md_ints, pdec_ints, fifo_ints, hact, vact;
 	bool plugin;
 	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
 
@@ -1263,42 +1448,56 @@ static int rk628_bt1120_isr(struct v4l2_subdev *sd, u32 status, bool *handled)
 		v4l2_err(sd, "handled NULL, err return!\n");
 		return -EINVAL;
 	}
-
-	rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_ISTS, &md_ints);
 	rk628_i2c_read(bt1120->rk628, HDMI_RX_PDEC_ISTS, &pdec_ints);
-	plugin = tx_5v_power_present(sd);
-	v4l2_dbg(1, debug, sd, "%s: md_ints: %#x, pdec_ints:%#x, plugin: %d\n",
-		__func__, md_ints, pdec_ints, plugin);
-
-	if ((md_ints & (VACT_LIN_ISTS | HACT_PIX_ISTS |
-			HS_CLK_ISTS | DE_ACTIVITY_ISTS |
-			VS_ACT_ISTS | HS_ACT_ISTS))
-			&& plugin) {
-		rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_HACT_PX, &hact);
-		rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_VAL, &vact);
-		v4l2_dbg(1, debug, sd, "%s: HACT:%#x, VACT:%#x\n",
-				__func__, hact, vact);
-
-		rk628_bt1120_enable_interrupts(sd, false);
-		enable_stream(sd, false);
-		bt1120->nosignal = true;
-		schedule_delayed_work(&bt1120->delayed_work_res_change, HZ / 2);
-
-		v4l2_dbg(1, debug, sd, "%s: hact/vact change, md_ints: %#x\n",
-				__func__, (u32)(md_ints & (VACT_LIN_ISTS |
-					HACT_PIX_ISTS)));
-		*handled = true;
+	if (bt1120->ctsn_ints_en) {
+		if (pdec_ints & (ACR_N_CHG_ICLR | ACR_CTS_CHG_ICLR)) {
+			rk628_csi_isr_ctsn(sd, pdec_ints);
+			pdec_ints &= ~(ACR_CTS_CHG_ICLR | ACR_CTS_CHG_ICLR);
+			*handled = true;
+		}
+	}
+	if (bt1120->fifo_ints_en) {
+		rk628_i2c_read(bt1120->rk628, HDMI_RX_AUD_FIFO_ISTS, &fifo_ints);
+		if (fifo_ints & 0x18) {
+			rk628_csi_isr_fifoints(sd, fifo_ints);
+			*handled = true;
+		}
 	}
+	if (bt1120->vid_ints_en) {
+		rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_ISTS, &md_ints);
+		plugin = tx_5v_power_present(sd);
+		v4l2_dbg(1, debug, sd, "%s: md_ints: %#x, pdec_ints:%#x, plugin: %d\n",
+			 __func__, md_ints, pdec_ints, plugin);
+
+		if ((md_ints & (VACT_LIN_ISTS | HACT_PIX_ISTS |
+				HS_CLK_ISTS | DE_ACTIVITY_ISTS |
+				VS_ACT_ISTS | HS_ACT_ISTS))
+				&& plugin) {
+
+			rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_HACT_PX, &hact);
+			rk628_i2c_read(bt1120->rk628, HDMI_RX_MD_VAL, &vact);
+			v4l2_dbg(1, debug, sd, "%s: HACT:%#x, VACT:%#x\n",
+				 __func__, hact, vact);
+
+			rk628_bt1120_enable_interrupts(sd, false);
+			enable_stream(sd, false);
+			bt1120->nosignal = true;
+			schedule_delayed_work(&bt1120->delayed_work_res_change, HZ / 2);
+
+			v4l2_dbg(1, debug, sd, "%s: hact/vact change, md_ints: %#x\n",
+				 __func__, (u32)(md_ints & (VACT_LIN_ISTS | HACT_PIX_ISTS)));
+			*handled = true;
+		}
 
-	if ((pdec_ints & AVI_RCV_ISTS) && plugin) {
-		v4l2_dbg(1, debug, sd, "%s: AVI RCV INT!\n", __func__);
-		bt1120->avi_rcv_rdy = true;
-		/* After get the AVI_RCV interrupt state, disable interrupt. */
-		rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_CLR, 0xffffffff);
+		if ((pdec_ints & AVI_RCV_ISTS) && plugin) {
+			v4l2_dbg(1, debug, sd, "%s: AVI RCV INT!\n", __func__);
+			bt1120->avi_rcv_rdy = true;
+			/* After get the AVI_RCV interrupt state, disable interrupt. */
+			rk628_i2c_write(bt1120->rk628, HDMI_RX_PDEC_IEN_CLR, AVI_RCV_ISTS);
 
-		*handled = true;
+			*handled = true;
+		}
 	}
-
 	if (*handled != true)
 		v4l2_dbg(1, debug, sd, "%s: unhandled interrupt!\n", __func__);
 
@@ -2178,8 +2377,10 @@ static int rk628_bt1120_probe(struct i2c_client *client,
 			rk628_bt1120_delayed_work_enable_hotplug);
 	INIT_DELAYED_WORK(&bt1120->delayed_work_res_change,
 			rk628_delayed_work_res_change);
+	INIT_DELAYED_WORK(&bt1120->delayed_work_audio_rate_change,
+			rk628_csi_delayed_work_audio_rate_change);
 	INIT_DELAYED_WORK(&bt1120->delayed_work_audio,
-			rk628_bt1120_delayed_work_audio);
+			rk628_csi_delayed_work_audio);
 	rk628_bt1120_initial_setup(sd);
 
 	if (bt1120->hdmirx_irq) {
@@ -2223,7 +2424,6 @@ static int rk628_bt1120_probe(struct i2c_client *client,
 		goto err_work_queues;
 	}
 
-	rk628_bt1120_enable_interrupts(sd, tx_5v_power_present(sd));
 	err = v4l2_ctrl_handler_setup(sd->ctrl_handler);
 	if (err) {
 		v4l2_err(sd, "v4l2 ctrl handler setup failed! err:%d\n", err);
@@ -2241,6 +2441,7 @@ err_work_queues:
 	cancel_delayed_work(&bt1120->delayed_work_enable_hotplug);
 	cancel_delayed_work(&bt1120->delayed_work_res_change);
 	cancel_delayed_work(&bt1120->delayed_work_audio);
+	cancel_delayed_work(&bt1120->delayed_work_audio_rate_change);
 err_hdl:
 	mutex_destroy(&bt1120->confctl_mutex);
 	media_entity_cleanup(&sd->entity);
-- 
2.17.1

